rcl/src/rcl/publisher.c

```diff
@@ -421,7 +421,11 @@ rcl_publisher_is_valid(const rcl_publisher_t * publisher)
     return false;  // error already set
   }
   if (!rcl_context_is_valid(publisher->impl->context)) {
-    RCL_SET_ERROR_MSG("publisher's context is invalid");
+    if (!rcl_error_is_set()) {
+      // rcl_context_is_valid can return false both in the error case, and when the context
+      // hasn't been initialized.  It will only set the error message in the first case.
+      RCL_SET_ERROR_MSG("publisher's context is invalid");
+    }
```

Comment:
i would do the following only to use the designed interfaces. (relying on the implementation behavior could be changed, for example we could add the error message if the context id is zero in the future? that breaks this if statement.)

```suggestion
    RCL_SET_ERROR_MSG("publisher's context is either invalid or uninitialized");
```

Reply:
Unfortunately, we can't do this and still get rid of the warning.  That's because `rcl_context_is_valid` can set an error if, for example, the `context` argument is NULL.  In that case, it will return with `false`, but also have set the error message, so setting it again here leads to the "overwritten" warning message.  Unfortunately, `rcl_context_is_valid` can *also* return `false` in the case where the context has already been shutdown.  In that case, it does *not* set the error, and returns `false`.  In both cases we want to fail here, but we don't necessarily want to overwrite the error message.  That's why we have this funky code.

(I'll also note that this is what https://github.com/ros2/rcl/pull/1086 was trying to fix, but that was too difficult to do)

Reply:
thanks for the explanation.